salt/returners/prometheus_textfile.py

```diff
@@ -0,0 +1,332 @@
+            "help": "Number of salt minion processes running",
+            "value": salt_procs,
+        },
+        "salt_states_succeeded": {
+            "help": "Number of successful states in the run",
+            "value": success,
+        },
+        "salt_states_failed": {
+            "help": "Number of failed states in the run",
+            "value": failure,
+        },
+        "salt_states_changed": {
+            "help": "Number of changed states in the run",
+            "value": changed,
+        },
+        "salt_states_total": {
+            "help": "Total states in the run",
+            "value": total,
+        },
+        "salt_states_success_pct": {
+            "help": "Percent of successful states in the run",
+            "value": round((success / total) * 100, 2),
+        },
+        "salt_states_failure_pct": {
+            "help": "Percent of failed states in the run",
+            "value": round((failure / total) * 100, 2),
+        },
+        "salt_states_changed_pct": {
+            "help": "Percent of changed states in the run",
+            "value": round((changed / total) * 100, 2),
+        },
+        "salt_elapsed_time": {
+            "help": "Time spent for all operations during the state run",
+            "value": round(duration, 3),
+        },
+        "salt_last_started": {
+            "help": "Estimated time the state run started",
+            "value": int(now - duration / 1000),
+        },
+        "salt_last_completed": {
+            "help": "Time of last state run completion",
+            "value": now,
+        },
+    }
+
+    if opts["add_state_name"]:
+        old_name = os.path.basename(opts["filename"])
+        new_name = (
+            "".join(old_name.split(".")[:-1])
+            + "-{}.".format(prom_state)
+            + "".join(old_name.split(".")[-1])
+        )
+        opts["filename"] = os.path.join(out_dir, new_name)
+        log.debug("Modified Prometheus filename from %s to %s", old_name, new_name)
+        for key in list(output.keys()):
+            output[key + '{state="' + prom_state + '"}'] = output.pop(key)
+
+    if opts["mode"]:
+        try:
+            opts["mode"] = int(opts["mode"], base=8)
+        except (KeyError, ValueError):
+            opts["mode"] = None
+            log.exception("Unable to convert mode to octal. Using system default.")
+
+    try:
+        with salt.utils.files.fpopen(
+            opts["filename"],
+            "w",
+            uid=opts["uid"],
+            gid=opts["gid"],
+            mode=opts["mode"],
+        ) as textfile:
+            textfile.write(
+                "\n".join(
+                    [
+                        "# HELP {} {}\n# TYPE {} gauge\n{} {}".format(
+                            k.split("{")[0], v["help"], k.split("{")[0], k, v["value"]
+                        )
+                        for k, v in output.items()
+                    ]
+                )
+                + "\n"
+            )
```

Comment:
**suggestion** (non-blocking) this might be clearer written like this: (note - I have no clue what would be actual descriptive names here:

```
            for k,v in output.items():
                thing = k.split("{")[0]
                print(
                    "# HELP {thing} {v[help]}\n# TYPE {thing} gauge\n{k} {v[value]}".format(
                        k=k, v=v, thing=thing
                    ),
                    file=textfile,
                )
```

Or alternatively, using f-strings:

```
            for k, v in output.items:
                thing = k.split("{")[0]
                print(
                    f"# HELP {thing} {v['help']}\n# TYPE {thing} gauge\n{k} {v['value']}",
                    file=textfile2,
                )
```